src/HazelcastClient.ts

```diff
@@ -370,64 +338,107 @@ export default class HazelcastClient {
     }
 
     /**
-     * Shuts down this client instance.
+     * Returns the {@link AddressProvider} of the client.
      */
-    shutdown(): void {
-        this.lifecycleService.emitLifecycleEvent(LifecycleEvent.shuttingDown);
+    getAddressProvider(): AddressProvider {
+        return this.addressProvider;
+    }
+
+    getLoadBalancer(): LoadBalancer {
+        return this.loadBalancer;
+    }
+
+    doShutdown(): void {
         if (this.mapRepairingTask !== undefined) {
             this.mapRepairingTask.shutdown();
         }
         this.nearCacheManager.destroyAllNearCaches();
-        this.statistics.stop();
-        this.partitionService.shutdown();
-        this.heartbeat.cancel();
+        this.proxyManager.destroy();
         this.connectionManager.shutdown();
-        this.listenerService.shutdown();
         this.invocationService.shutdown();
-        this.lifecycleService.emitLifecycleEvent(LifecycleEvent.shutdown);
+        this.listenerService.shutdown();
+        this.statistics.stop();
+    }
+
+    /**
+     * Shuts down this client instance.
+     */
+    shutdown(): void {
+        this.getLifecycleService().shutdown();
```

Comment:
Shouldn't this method be merged with `doShutdown()`?

Reply:
The logic is okay but we are exposing the `doShutdown` method. In the Java client `doShutdown` is in the implementation, not on the public API. So, I think we could mark them as private and access them using `as any` trick. Would that be okay ? This is the same problem as `sendStateToCluster`

Reply:
I see. I was confused that the client calls `LifecycleService` which is in its turn calls the client. It would be nice if we could simplify this somehow.

As for the `as any`, let's avoid using it for now. In future we can change visibility of `doShutdown()` as it's not documented and we don't have any compatibility obligations for it.